File: 20-ValidParentheses/20-ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> myStack;
        for (int i = 0; i < s.length(); i++){
            if (s[i] == '{' || s[i] == '(' || s[i] == '['){
                myStack.push(s[i]);
            }
            else if (s[i] == ')'){
                if (myStack.empty() || myStack.top() != '('){
                    return false;
                }
                else {
                    myStack.pop();
                }

            }
            else if (s[i] == '}'){
                if (myStack.empty() || myStack.top() != '{'){
                    return false;
            }
            else {
                    myStack.pop();
                }

            }
            else if (s[i] == ']'){
                if (myStack.empty() || myStack.top() != '['){
                    return false;
            }
            else {
                    myStack.pop();
                }

            }
            
        }
        if (myStack.empty()){
        return true;}

        return false;
    }
};

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 7/2/2026, 11:17:58 PM
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        alist = []

        for char in s:
            if char in '({[':
                alist.append(char)
            elif char == ']' and (not alist or alist[-1] != '['):
                return False
            elif char == '}' and (not alist or alist[-1] != '{'):
                return False
            elif char == ')' and (not alist or alist[-1] != '('):
                return False  
            else:
                alist.pop()

        return not alist